src/string_bytes.cc

```diff
@@ -656,20 +678,108 @@ static void force_ascii(const char* src, char* dst, size_t len) {
 }
 
 
-static size_t hex_encode(const char* src, size_t slen, char* dst, size_t dlen) {
+#define pack(byte) (static_cast<uint8_t>(byte) * 0x01010101u)
+uint32_t nibbles_to_hex(uint32_t nibbles) {
+  const uint32_t ascii09 = nibbles + pack('0');
+  const uint32_t correction = pack('a' - '0' - 10);
+
+  const uint32_t tmp = nibbles + pack(128 - 10);
+  const uint32_t msb = tmp & pack(0x80);
+  const uint32_t mask = msb - (msb >> 7);
+
+  return ascii09 + (mask & correction);
+}
+
+
+#define byte_to_hex(src, dst, i, val)                                          \
+  do {                                                                         \
+    val = static_cast<uint8_t>(src[i]);                                        \
+    dst[i * 2] = hex[val >> 4];                                                \
+    dst[i * 2 + 1] = hex[val & 15];                                            \
+    if (slen == i + 1) return;                                                 \
+  } while (0)
+
+static void hex_encode(const char* src, size_t slen, char* dst, size_t dlen) {
   // We know how much we'll write, just make sure that there's space.
   CHECK(dlen >= slen * 2 &&
       "not enough space provided for hex encode");
 
-  dlen = slen * 2;
-  for (uint32_t i = 0, k = 0; k < dlen; i += 1, k += 2) {
+  // Process smaller inputs first
+  {
+    if (slen == 0) return;
     static const char hex[] = "0123456789abcdef";
-    uint8_t val = static_cast<uint8_t>(src[i]);
-    dst[k + 0] = hex[val >> 4];
-    dst[k + 1] = hex[val & 15];
+    uint8_t val;
+    byte_to_hex(src, dst, 0, val);
+    byte_to_hex(src, dst, 1, val);
+    byte_to_hex(src, dst, 2, val);
+    byte_to_hex(src, dst, 3, val);
+    byte_to_hex(src, dst, 4, val);
+    byte_to_hex(src, dst, 5, val);
+    byte_to_hex(src, dst, 6, val);
+    byte_to_hex(src, dst, 7, val);
+    byte_to_hex(src, dst, 8, val);
+    byte_to_hex(src, dst, 9, val);
+    byte_to_hex(src, dst, 10, val);
+    byte_to_hex(src, dst, 11, val);
+    byte_to_hex(src, dst, 12, val);
+    byte_to_hex(src, dst, 13, val);
+    byte_to_hex(src, dst, 14, val);
+    byte_to_hex(src, dst, 15, val);
+  }
+
+  size_t i = 0;
+  size_t d = 32;
+  size_t nb = 16;
+  uint16_t* src16 = reinterpret_cast<uint16_t*>(const_cast<char*>(src + 16));
+  size_t alignment = reinterpret_cast<uintptr_t>(src16) % sizeof(*src16);
```

Comment:
This isn't reliable, you need to do it before casting the `char*` to `uint16_t*`.